lib/elixir/lib/enum.ex

```diff
@@ -2734,11 +2734,15 @@ defmodule Enum do
       [{1, :a}, {2, :b}, {3, :c}]
 
   """
-  @spec zip([t]) :: t
+  @spec zip([t] | %Stream{}) :: t
 
   def zip([]), do: []
 
-  def zip(enumerables) when is_list(enumerables) do
+  def zip(%{__struct__: Stream} = enumerables), do: do_zip(enumerables)
```

Comment:
Used pattern matching for the `__struct__` key because clean compilation fails if I use this:

```
def zip(%Stream{} = enumerables), do: do_zip(enumerables)
```
I'm wondering if we need to check the `enumerable` type here if `Stream.zip/1` already does it.